extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-preview.js

```diff
@@ -552,19 +555,45 @@ export class AmpStoryDevToolsTabPreview extends AMP.BaseElement {
     }).then(() => {
       deviceSpecs.player
         .getElement()
-        .addEventListener('storyNavigation', (event) => {
-          this.devices_.forEach((d) => {
-            if (d != deviceSpecs) {
-              d.player.show(null, event.detail.pageId);
-            }
-          });
-        });
+        .addEventListener('storyNavigation', (event) =>
+          this.onPlayerNavigation_(event, deviceSpecs)
+        );
       deviceSpecs.player.load();
     });
+    this.expectedNavigationEvents_[deviceSpecs.name] = [];
     this.devices_.push(deviceSpecs);
     this.updateDevicesInHash_();
   }
 
+  /**
+   * Triggered when a player emits a storyNavigationEvent.
+   *
+   * A navigation event from a player can come from a user interaction or a previous programmatic call.
+   * Expected navigation events from programmatic calls are stored in `this.expectedNavigationEvents_`,
+   * so they should not be propagated (but deleted from the list of expected events). If an event was not
+   * expected, it means it was user navigation and should be propagated to other players.
+   * @param {!Event} event
+   * @param {!DeviceInfo} deviceSpecs
+   * @private
+   */
+  onPlayerNavigation_(event, deviceSpecs) {
+    const {pageId} = event.detail;
+    const pageIndex = this.expectedNavigationEvents_[
+      deviceSpecs.name
+    ].lastIndexOf(pageId);
+    if (pageIndex > -1) {
+      // Remove the expected events up to the most recently received event if it was in the list.
+      this.expectedNavigationEvents_[deviceSpecs.name].splice(0, pageIndex + 1);
```

Comment:
I must be missing something, but IIUC there could only be one event at most in the `this.expectedNavigationEvents_[device]` whenever we get to this point, right? If so, could we simplify this a bit like so?:

```javascript
if (this.expectedNavigationEvents_[d.name].indexOf(pageId) != -1) {
  this.expectedNavigationEvents_[d.name].pop();
  return;
}
// ...
this.expectedNavigationEvents_[d.name].push(pageId);
```

or even 
```javascript
if (this.expectedNavigationEvent_[d.name]) {
  this.expectedNavigationEvent_[d.name] = false;
  return;
}
// ...
d.player.show(..)
this.expectedNavigationEvent_[d.name] = true;
```

Reply:
The problem is that there's a chance that you send an event while the story is loading and you'd have an unmatched pageId that stays there forever (and would cause problems if you navigate to that page), so when we use this list as a queue of events (and remove the previous ones that were unmatched), then we fix that error before it becomes a problem.

Also you could have 2 or more expected navigation events if users click fast and the players take a bit to navigate

Reply:
Ok, could you add a comment explaining these two cases? A unit test for this would be useful too.

Reply:
Yeah I'll add a comment for it. I'm not sure how hard it will be to write/run a test that loads several players